C and Fortran API for Interpolator

Fuad-HH opened the pull request:
- [x] API
- [x] Test API
- [x] More parameters for controlling interpolator behavior
- [x] Remove hardcoded parameters and prints
- [x] Fix #177 
- [x] Fix `test_normalization_routine`
- [x] Fix other interpolation tests
- [x] Fix #207 
- [x] Fix the test_interpolation_on_ltx_mesh
- [x] Virtual destructor for interpolator base
- [x] Merge APIs for Point Cloud-based and MLS
- [x] Add file and method documentation
- [x] Merge omega_h handle pointer

Fuad-HH committed:
interpolator base class, tests [breaking]

the test is still not passing. pushed it to save it at current state

Fuad-HH committed:
working single mesh interpolation class

the values are not being compared yet

Fuad-HH committed:
interpolator test compares each entry with expected

passing with 10% error margin. [future] find best degree and min required support to minimize the epsilon of the test

Fuad-HH committed:
interpolator test passes for mesh to mesh

with 10% error margin and [temporary] it doesn't work when a source and a target becomes coincident: discussed here: https://github.com/SCOREC/pcms/pull/148#discussion_r1926204199

Fuad-HH committed:
adding c api

compiling, yet to be properly merged to the other apis and testing

Fuad-HH committed:
add omega_h mesh handle struct

Fuad-HH committed:
fortran api for interpolator

- compiling
- yet to be completed and tested

Fuad-HH committed:
fortran and c api to read Omega_h mesh from file

keeps the mesh and lib alive

Fuad-HH committed:
added test_interpolator_fortran

- test is going through and finding the supports but goes infinitely
- needs number of supports and max_radius etc. as parameters
- still needs more structs, enums to represent interpolation types and other
  functionalities
- still needs evaluate function

Fuad-HH committed:
interpolator frotran api test correction

now the test is passing and finding supports

- the mesh is used temporarily

Fuad-HH committed:
added pcms_interpolate function

Fuad-HH committed:
test pcms interpolator api

test is passing for an uniform field. It just checks if the
functionalities are at the place

Fuad-HH committed:
point cloud based search working

- tested but failing for some specific target points

Fuad-HH committed:
point cloud search

Fuad-HH committed:
merge c and fortran api with develop

Fuad-HH committed:
mesh based and mixed interpolator for degas2 xgc coupling

when xgc's corresponding omega_h mesh is not available,it needs
mixed interpolation.
all these methods are based on point cloud
they can be improved
search can be unstable (find better ways) and
changed the radius adaptation so that it stops after 100
iterations even if it hasn't found enough supports

Fuad-HH committed:
minor parameter changes for better interpolation

Fuad-HH committed:
Merge branch 'mergecfortapi' of github.com:Fuad-HH/pcms into mergecfortapi

Fuad-HH committed:
get number of mesh elements from interpolation creating call

jacobmerson commented on src/pcms/capi/interpolator.cpp:
```diff
@@ -27,6 +27,73 @@ PcmsPointBasedInterpolatorHandle pcms_create_point_based_interpolator(void* sour
+  dg2_fname = dg2_fname.erase(dg2_fname.find_last_not_of(" \n\r\t")+1);
+  printf("The interpolator got xgc mesh file: %s and dg2 mesh file: %s\n", xgc_fname.c_str(), dg2_fname.c_str());
+
+  // read the meshes
```
This may be a question for @cwsmith if Omega_h supports having two separate meshes in parallel in the same library.

jacobmerson commented on src/pcms/interpolator/interpolation_base.cpp:
```diff
@@ -153,6 +153,8 @@ void MLSPointCloudInterpolation::find_supports(uint min_req_supports)
   uint min_supports_found = 0;
   uint max_supports_found = 0;
   // radius adjustment loop
+  int loop_count = 0;
+  int max_count = 100;
```
Sounds good. It's marked as draft anyways I just wanted to provide some feedback while I had time.

Fuad-HH committed:
add xgc node to degas2 centroid interpolation

Fuad-HH committed:
merge pcms interpolator and its c/fortran API with develop

Fuad-HH committed:
fix merging error with interpolator normalization

Fuad-HH committed:
fix interpolation class test partially (the values)

still needs to solve the problem with number of neighbors in mesh based search by merging the both radius adaptation

Fuad-HH committed:
divided point cloud support strcuture search

remains todos

Fuad-HH committed:
adj searches use helper functions to adapt radius

separate helper function header
passes interpolation class test
fix issue #207

Fuad-HH committed:
clang-format on the changed source files

Fuad-HH committed:
replace prints with pcms::printInfo or Error

except for prints inside gpu function since fmt is not
supported for GPU

Fuad-HH committed:
interpolation parameters exposed to fortran api

Fuad-HH committed:
separate mpi comm for all mesh readers

jacobmerson commented on src/pcms/capi/interpolator.h:
```diff
@@ -0,0 +1,72 @@
+PcmsPointBasedInterpolatorHandle pcms_create_xgcnodedegas2_interpolator(
+  const char* dg2_mesh_filename, void* source_points, int source_points_size,
+  double radius, void* dg2_elem_count, int degree, int min_req_supports,
+  double lambda, double decay_factor);
```
I don't think most of the internal interpolation routines should be specific to degas2/xgc. I'd rather that we have a reader that creates a field handle for xgc and degas2 that can be directly used with the more general interpolation APIs.

jacobmerson commented on src/pcms/interpolator/interpolation_base.cpp:
```diff
@@ -0,0 +1,463 @@
+        factor = (factor < 0.1) ? 0.33 : factor;
+        radii2_l[i] *= factor;
+      }
+      num_supports[i] = 0; // reset the support pointer
```
I'm not sure that I have a good solution here in terms of how to pick that factor for searching. @MohammadElahiNU did implement the uniform grid search with a circle/sphere that should increase the speed of the search for points only in https://github.com/SCOREC/pcms/pull/176 .

However, that pull request never got merged because it had a bunch of duplicated code that needs to get cleaned up. Mohammad is mostly working on other stuff and we didn't have a chance to get back to it.

jacobmerson commented on src/pcms/capi/CMakeLists.txt:
```diff
@@ -19,8 +19,22 @@ set_target_properties(
+set(INT_CAPI_HEADERS ${CMAKE_CURRENT_SOURCE_DIR}/interpolator.h)
+set_target_properties(pcms_interpolator_capi
+  PROPERTIES PUBLIC_HEADERS "${INT_CAPI_HEADERS}"
+             OUTPUTNAME pcmsinterpolatorcapi
```
Indeed, `OUTPUT_NAME` is correct: https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#target-properties

jacobmerson commented on src/pcms/capi/interpolator.cpp:
```diff
@@ -0,0 +1,241 @@
+//
+#include <pcms/capi/kokkos.h>
+#include "interpolator.h"
+#include <interpolation_base.h>
```
use full path

jacobmerson commented on src/pcms/capi/interpolator.cpp:
```diff
@@ -0,0 +1,241 @@
+  }
+}
+
+PcmsPointBasedInterpolatorHandle pcms_create_degas2xgcnode_interpolator(
```
is this a case we currently need to support?

Fuad-HH replied:
Yes, I use this because in Degas2 XGC coupling, we need to transfer data from Degas2 centroids XGC nodes. Used [here](https://github.com/Fuad-HH/XGC-Devel/blob/8d910ec8b0f850a96e9c72e546b2dc22116ef572/XGC_core/neutral_totalf.F90#L895).

jacobmerson commented on src/pcms/interpolator/interpolation_base.h:
```diff
@@ -0,0 +1,142 @@
+/**
+ * @brief Moving Least Square Radial Basis Function Interpolation
+ */
+class MLSInterpolationHandler : public InterpolationBase
```
There may be an opportunity to join with `MLSPointCloudInterpolator`? It seems like most things besides the search should be the same.

If not, we need to rename this. I don't know what `Handler` means. Maybe `MLSMeshInterpolator`?

jacobmerson commented on src/pcms/interpolator/interpolation_base.h:
```diff
@@ -0,0 +1,142 @@
+  Omega_h::HostWrite<Omega_h::Real> target_field_;
+  Omega_h::HostWrite<Omega_h::Real> source_field_;
+
+  void find_supports(uint min_req_supports = 10);
```
Maybe a baseclass for the search would help us combine things?

jacobmerson commented on src/pcms/interpolator/interpolation_base.cpp:
```diff
@@ -0,0 +1,463 @@
+void MLSPointCloudInterpolation::distance_based_pointcloud_search(
+  Omega_h::Write<Omega_h::Real> radii2_l,
+  Omega_h::Write<Omega_h::LO> num_supports) const
+{
```
Add note to use search structure once https://github.com/SCOREC/pcms/pull/176 is merged.

jacobmerson commented on src/pcms/interpolator/interpolation_base.cpp:
```diff
@@ -0,0 +1,463 @@
+  OMEGA_H_CHECK_PRINTF(source_field.size() == source_coords_.size() / dim_,
+                       "Target Data and Target Points size mismatch: %zu %d\n",
+                       source_field.size(), source_coords_.size() / dim_);
+  for (int i = 0; i < 10; i++) {
```
this looks like some debug printing. I think it should be removed, or at least only printed in a debug mode.

Fuad-HH committed:
remove assertion skip

Fuad-HH commented on src/pcms/capi/interpolator.h:
```diff
@@ -0,0 +1,72 @@
+};
+typedef struct PcmsInterpolatorHandle PcmsInterpolatorHandle;
+
+struct PcmsPointBasedInterpolatorHandle
```
Yes, I agree. I could not find a better way to handle this. I am considering using an inherited class with a better virtual class than I currently have. I can work on it.

Fuad-HH committed:
add printDebugInfo and fix typo

Fuad-HH commented on src/pcms/interpolator/interpolation_base.cpp:
```diff
@@ -0,0 +1,463 @@
+    "Size mismatch in copy_data_from_HostWrite_to_ScalarArray: %d %zu\n",
+    source.size(), target.size());
+
+  for (int i = 0; i < source.size(); ++i) {
```
Checking with `data_handle` and `data` methods from PCMS and Kokkos, respectively.

Fuad-HH committed:
check pointers not pointing same array when copying

Fuad-HH commented on src/pcms/print.h:
```diff
@@ -43,6 +43,22 @@ KOKKOS_INLINE_FUNCTION void printInfo(const char* fmt, const Args&... args)
+#endif
+#endif
+}
+
```
I needed it on the GPU too, so I made it a little different than the others here. I think those should have a similar print statement for the GPU, too.

Fuad-HH committed:
add test case to test a real example of interpolation

This is from XGC-Degas2 coupling

It also modifies the interpolator constructor to allow
any kind of point values including const and nonconst

Fuad-HH committed:
better parameters

still the max error is about 30% at couple of nodes near the boundary

Fuad-HH committed:
add virtual destructor

Fuad-HH committed:
rename MLSMeshInterpolation

Fuad-HH committed:
Unify interpolator handles in C/Fortran API

- Replace all PcmsPointBasedInterpolatorHandle with PcmsInterpolatorHandle
- Consolidate destroy and interpolate functions into generic InterpolatorHandle.
- Added tests to use unified interpolator API and add Fortran interpolation test.

Fuad-HH committed:
refactor ltx interpolation test, improved verification

- two way interpolation using both veriables
- check l2 norm
- improved error handling

Fuad-HH committed:
merge develop

Fuad-HH commented:
The above commit (https://github.com/SCOREC/pcms/pull/178/commits/4013be7f55701e713517412169aa4ae95b98e681) is failing two tests:
```bash
      Start 15: interpolate linear 2d omega_h_field
15/44 Test #15: interpolate linear 2d omega_h_field ......Subprocess aborted***Exception:   1.98 sec
      Start 16: interpolate quadratic 2d omega_h_field
16/44 Test #16: interpolate quadratic 2d omega_h_field ...Subprocess aborted***Exception:   4.89 sec
```
Although the tests are not failing in the CI/CD.

jacobmerson commented:
It looks like the test failures that are coming up is related to: https://github.com/SCOREC/meshFields/issues/69

Fuad-HH committed:
remove MPI_Comm_dup for omega_h reads

The given communicator is already being duplicated by the library
constructor here: https://github.com/SCOREC/omega_h/blob/e8fe57b127eff97316735ca75678eb3f66ad72d0/src/Omega_h_library.cpp#L92

However, this assumes that MPI is already initialized and Omega_h is
built with OMEGA_H_USE_MPI

Fuad-HH committed:
minor refactorings

- remove PCMS_ENABLE_PRINT from printDebugInfo
It is debug info and should not depend on the flag and should always
print

- max allowed neighbors brought out on surface level APIs
- minor print format fixes

Fuad-HH commented on src/pcms/print.h:
```diff
@@ -43,6 +43,22 @@ KOKKOS_INLINE_FUNCTION void printInfo(const char* fmt, const Args&... args)
+  printf(fmt, args...);
+#endif
+#endif
+}
```
Removed ` && defined(PCMS_PRINT_ENABLED)` so that it prints in https://github.com/SCOREC/pcms/pull/178/commits/ceb6334cb956f67c32e6e55a3cff29e9e13825b5. Should not depend on `PCMS_PRINT_ENABLED` flag.

Fuad-HH committed:
fix node file read

Fuad-HH commented on test/test_interpolation_on_ltx_mesh.cpp:
```diff
@@ -0,0 +1,385 @@
+        printf("[WARNING] Non-zero Z coordinate found (%f) for node ID %d. "
+               "Expected Z=0 for 2D mesh.\n",
+               z, node_id);
+      }
```
Removed in https://github.com/SCOREC/pcms/pull/178/commits/ce865846f75722cf94022118f748005b2228c1a6.

Fuad-HH committed:
make decay_factor and lambda parameters

Fuad-HH committed:
remove degas2xgc c/fortran API

Since XGC won't have osh file in real life, it is pointless for now.

Fuad-HH commented on test/test_interpolation_class.cpp:
```diff
@@ -0,0 +1,330 @@
+      REQUIRE_THAT(
+        point_cloud_interpolated_data_hwrite[i],
+        Catch::Matchers::WithinAbs(interpolated_data_hwrite[i], 1e-4));
+    }
```
It's fine. It can be helpful.

Fuad-HH committed:
replace mesh in interpolation test

any small, simple, 2D mesh would work, already available mesh in pcms_testcase repository makes more sense.

Fuad-HH committed:
temporary fix for ltx interpolation test

Fuad-HH committed:
add PCMS_PRINT_ENABLED for printDebugInfo

To be in accordance with xsdk

Fuad-HH committed:
add doc to interpolation_base.h

Fuad-HH committed:
add doc to interpolator.h

Fuad-HH committed:
temporary fix to print fail log

Fuad-HH committed:
always print log

Fuad-HH committed:
Revert "always print log"

This reverts commit b10cfb19a19e991002c96b361dabd938bcf1de15.

Fuad-HH committed:
Revert "temporary fix to print fail log"

This reverts commit 35d20d4a4c73dc2996b5c0d0e0ce88c5250c6d17.

jacobmerson commented on src/pcms/fortranapi/interpolator.i:
```diff
@@ -0,0 +1,39 @@
+%include <stdint.i>
+%include <typemaps.i>
+
+struct PcmsInterpolatorOHMeshHandle
```
Is there a reason we need a different OmegaH handle than the one that's in the client.h?

Fuad-HH replied:
I was trying to write a test case to demonstrate that if the library goes out of scope before the mesh, the mesh remains intact or not; it looks like it remains intact. I am not sure what's going on there.

jacobmerson commented on src/pcms/capi/interpolator.cpp:
```diff
@@ -0,0 +1,155 @@
+  }
+}
+
+PcmsInterpolatorOHMeshHandle read_oh_mesh(const char* filename)
```
rename to `pcms_create_omega_h_mesh`

jacobmerson commented on src/pcms/capi/interpolator.cpp:
```diff
@@ -0,0 +1,155 @@
+  return {reinterpret_cast<void*>(mesh), reinterpret_cast<void*>(oh_lib)};
+}
+
+void release_oh_mesh(PcmsInterpolatorOHMeshHandle oh_mesh)
```
`destroy_omega_h_mesh`

Fuad-HH committed:
rename omega_h mesh handle api

Fuad-HH committed:
divide up mesh and library handle

Fuad-HH committed:
separate header for mesh

keeping it in the interpolator library object since it is only used in the interpolator.

Fuad-HH commented:
@jacobmerson, could you please review the changes I made to separate the mesh API to a different header?

However, I could not remove the mesh handle we already have in `client.h`.
I was trying to delete `PcmsOmegaHMeshHandle` and run swig after it. But the `client_wrap.c` file pcms currently has is not up to date with the `client.i` it has there. If I run swig on `client.i`, it changes another definition in the generated `.c` and `.f90` files. This is because the `PcmsClientHandle` has only a member `pointer` in `client.i`:
https://github.com/SCOREC/pcms/blob/a07efb906fa3421acc4b9be4dbfaaa67b8c537ee/src/pcms/fortranapi/client.i#L31-L34
But in `client.h`, it is defined as below
https://github.com/SCOREC/pcms/blob/a07efb906fa3421acc4b9be4dbfaaa67b8c537ee/src/pcms/capi/client.h#L10-L14


And it is producing the following error as expected
```bash
pcms/src/pcms/fortranapi/client_wrap.c:306:21: error: ‘struct PcmsClientHandle’ has no member named ‘pointer’; did you mean ‘appPointer’?
  306 |   if (arg1) (arg1)->pointer = arg2;
      |                     ^~~~~~~
      |                     appPointer
``` 

I think it is better to solve it in a different pull request. I can open an issue for it and solve it pretty quickly later.

jacobmerson commented on src/pcms/capi/mesh.h:
```diff
@@ -0,0 +1,82 @@
+extern "C" {
+#endif
+
+struct PcmsInterpolatorOmega_hLibraryHandle
```
This should be called `PcmsOmegaHLibraryHandle` to have consistent naming.

jacobmerson commented on src/pcms/capi/mesh.h:
```diff
@@ -0,0 +1,82 @@
+ * @brief Holds void pointers of Omega_h Mesh object and the library
+ * (communicator) handle
+ */
+struct PcmsInterpolatorOmega_hMeshHandle
```
This should be called PcmsOmegaHMeshHandle to have consistent naming.

jacobmerson reviewed (changes requested):
One more pedantic change to make the naming scheme consistent, but otherwise I'm ready to merge this.

Fuad-HH committed:
rename mesh and library handle

jacobmerson committed:
uint->unsigned

uint is not a standard type designator and did not compile on llvm

jacobmerson committed:
formatting with clang-format

jacobmerson committed:
add specific versions to workflow for redev and meshfields

jacobmerson committed:
Merge branch 'develop' into mergecfortapi

jacobmerson committed:
Merge branch 'develop' into mergecfortapi

jacobmerson committed:
add additional debugging output in ltx testcase

jacobmerson committed:
make sure test log is always output

Fuad-HH commented:
Looks like there's an issue reading the mesh in the test. Looking at the `build-Omega_h` phase of this latest [Test-Build](https://github.com/SCOREC/pcms/actions/runs/20071984159/job/57576355944?pr=178), I found that it is using Omega_h commit https://github.com/SCOREC/omega_h/commit/fbe1cc131fb1b5ac840129ecd8bd7b42ab244000. This commit has an issue reading this specific mesh used in this test case. It was fixed on the next commit. Could you please use the latest commit of Omega_h?

jacobmerson committed:
use git has for omega-h version in workflow until

https://github.com/SCOREC/omega_h/issues/195 is resolved

Fuad-HH committed:
change export name

Fuad-HH committed:
add kokkoskernel hint

Fuad-HH committed:
make cmake verbose

Fuad-HH committed:
debug-output

jacobmerson committed:
make kokkoskernels private link

jacobmerson committed:
fix build so mls_interpolation.hpp does not need the impl included.

The goal of this PR is to separate the mls_interpolation.hpp and mls_interpolation_impl.hpp so that the implementation is not included in the header file.

jacobmerson committed:
add kokkos kernels link to unit test

jacobmerson committed:
add interpolator target to pcms::pcms

jacobmerson committed:
rework capi targets

this commit modifies the targets for the capi so that pcms::capi is an interface that includes the subcomponents. This matches more closely to the way that the main pcms::pcms library has been put together.

jacobmerson committed:
add debugging output to the installation test case

jacobmerson committed:
rework fortranapi targets

this commit modifies the targets for the fortranapi so that pcms::capi is an interface that includes the subcomponents. This matches more closely to the way that the main pcms::pcms library has been put together.

jacobmerson committed:
update interpolator targets name

jacobmerson committed:
use pcms::pcms in example build directory

jacobmerson commented:
@Sichao25 @cwsmith any ideas on why this might be failing? It's working locally by running the test installation (albeit with a different version of CMake).

Sichao25 commented:
Maybe we can try to put `include("${CMAKE_CURRENT_LIST_DIR}/pcms_interpolator-targets.cmake")` before capi targets in `config.cmake.in`?

jacobmerson committed:
reorder config file to move interpolator target before c/fortran apis
